### DISCUSSION TITLE
Code debugging and database migrations

### DISCUSSION THREAD

**Discussion by martin-braun** [directus/directus#13743] [Archive]
Hi there. :)

I like to support directus since I discovered and used it in a project around one year ago. A very small project (licensing backend server with administrative capabilities). I think directus is brilliant, especially since it is open-source and free. I'm facing a larger project soon and want to consider directus as backend again to bypass all the necessary work for building an administrative backend once again, my customer is very happy with it and I would have more time for the precious front-end.

However, in my experience directus was rather hard to develop for: 

- I use Visual Studio Code, but despite directus running on node, I was unable to debug my extensions, because I think directus injects the extension scripts on runtime and runs them somewhere else
- I'm sure it is possible to write TypeScript, but it was not worth to bother, I would like to change that, but there is no sample project and no typings
- The inability to step into the code with debugger and investigate variables and possible types made it very hard to learn how to do simple things in such extensions
- directus' database structure is setup in the UI, so there is no database migration (for own models) which makes it hard to incorporate it into a repository with growing models, although I could imagine that directus could support creating snapshots from those models which allow to upgrade/downgrade migrations

> Please note: With "database migrations" I'm not talking about the system tables of directus. I'm rather talking about own tables that might get changed, added or removed in the process lifetime of a product that uses directus as database handler. This on the equivalent of Artisans or EntityFrameworks "create migration" process to alter or change the layout of the database. Something that is done in directus by simply editing the database on the UI, but I want to tie the current layout of the database to a commit on the repository and need the ability to rollback to test older versions with compatible database layouts. [Strapi has a plugin for that](https://www.npmjs.com/package/strapi-plugin-migrations), but I really want directus to be a successor here.

Did any of those things changed in the time? Can I get happy with directus? What are my options to address those problems?

**Comment by benhaynes** [comment]
I just wanted to say think you for the kind words — we'll get you a response tomorrow!

**Comment by rijkvanzanten** [comment]
* Debugging

Directus is a node process that uses regular `require()` to inject the custom logic. Therefore, if you run Directus itself from the VS Code debugger, you should be able to debug the extensions as well 🤔 

* Write typescript where? In custom extensions? The [`extensions-sdk`](https://docs.directus.io/extensions/creating-extensions/) can also scaffold extensions in typescript iirc

* See pt 1 :) 

* There's a `schema snapshot` and `schema apply`  command you can use to generate and apply schema files that can be tracked 🙂

**Reply by martin-braun** [reply]
Thanks for the heads up, I will look into this and return to this discussion.

**Reply by kevincam3** [reply]
I use WebStorm any help with how to setup it up in Webstorm. I've tried as well with no sucess.

**Reply by martin-braun** [reply]
@rijkvanzanten It's been a while, I wish to pickup my old discussion as I work on a new project in which I picked Directus once more eventually. I want to share my discoveries to give feedback and help others finding this thread.

I found https://github.com/bcc-code/directus-schema-sync which seems to leverage `schema snapshot` and `schema apply` internally, but it also automate the process via environment variables and it also exports and imports data, so I can specify default users, roles, etc. I think it would be awesome to have something like this without the need of an extension.

In regards of TypeScript and debugging in general, I figured loading extensions the old fashioned way doesn't seem to work anymore. The docs are horrible loose on that detail. The docs however are rich on building new rich extensions, but Directus behavior of loading those extensions other than through the `package.json` is undocumented. I wanted to have an endpoint extension as part of my project (not as separate extension) and it took a lot of trial and error to figure out that I can simply add an extension (created with `npx create-directus-extension@latest`) into the `extensions` directory and when I build it, Directus will see the `dist` folder and load my extension (no more `/custom/<name>/` route, route is now `/<extension-name>`). 

I think this use case is clunky, because I have to hack a bit within the root package.json to auto-build my endpoint extension. The process is flawless for extern extensions that I install via `npm`, but not very straight-forward for embedded extensions.

Anyways, I felt I need to share my experience in that regard.

<img width="805" alt="image" src="https://github.com/directus/directus/assets/24259245/7c639fe3-71f1-45fb-ba17-9eb7716098f7">

Very cool to have this working eventually.

**Reply by flevi29** [reply]
Yeah, the docs aren't very detailed, I too struggled way more than I could've with the new extensions loading format. I was thinking about using the `bundle` extension type to simplify things, but my toilet paper has better explanations on its packaging than Directus docs on how to set up a `bundle` extension.

Also I hoped the new way of loading extensions would make things simpler, but no, they just made it harder, because now I need a `package.json` for each of my extensions, so in my build script I also need to write the `package.json` -s. Previously I have proposed to have an option of a JS file describing each extension and their path, as that would make bundling extensions a lot simpler, but that was ignored.

**Reply by martin-braun** [reply]
Here is something else that bothers me a lot. @benhaynes @rijkvanzanten I would just like to mention that the types of `ApiExtensionContext` still contain `any`:

```ts
type ApiExtensionContext = {
    services: any;
    database: Knex;
    env: Record<string, any>;
    logger: Logger;
    getSchema: (options?: {
        accountability?: Accountability;
        database?: Knex;
    }) => Promise<SchemaOverview>;
};
```

`services` provides functions to read or modify the database, exactly what is used in a custom API to interact with the database, and it's very annoying to not get suggestions, causing any issues to be runtime issues exclusively.

<img width="633" alt="image" src="https://github.com/directus/directus/assets/24259245/7d2e03d3-7503-464f-a262-a990e29e3e33">

I really wish the team could invest into improving types for the extensions. I'm considering building much more on top of Directus, but the current state really makes me enjoy it for small extensions only. It would be very painful to have something rich that juggles with `any` pretty much.

**Comment by flevi29** [comment]
UPDATE: Better do it like in the comment below, so you don't have to rely on unexported `node_modules` file: https://github.com/directus/directus/discussions/13743#discussioncomment-8197663

For anyone else stumbling here in search for a satisfactory answer on how to debug Directus API extensions:
The methods of doing so suggested by the maintainers and others aren't ideal, and are awkward at best in my opinion.

Instead you can just:
`scripts/start.js`:
```javascript
// your setup code here, for example environment setup ....

import "../node_modules/directus/cli.js"; /* (".." or whatever path leads you to where node_modules is located) */
```
Make any necessary setup smoother, more readable and dynamic.

Ideally this `cli.js` should be exported for this reason, but it's not.

**Reply by flevi29** [reply]
I do get some weird breaks in two initiated threads, this might be because of how `cli.js` works, or `tsx` and lastly perhaps WebStorm. But hey, it works and I can set up my env however I want this way.

**Reply by MichaelFBA** [reply]
Can you provide an example of your startup script?

**Reply by flevi29** [reply]
Apparently the cli version is only needed if you also want Directus to check if it's out of date. That's not necessary, so we import `@directus/api/server`.

Now I'm doing it like so:
`start.ts`
```typescript
import { env as processEnv, default as process } from "node:process";
import { fileURLToPath, URL } from "node:url";
import { access } from "node:fs/promises";
import { env } from "./directus.config.js";

const extensionsDir = fileURLToPath(new URL("extensions", import.meta.url));

try {
  await access(extensionsDir);
} catch (error: unknown) {
  process.exitCode = 1;
  throw error;
}

env.EXTENSIONS_PATH = extensionsDir;

Object.assign(processEnv, env);

// Everything Directus must be imported after resolving env, because as soon
// as anything Directus related is imported, Directus reads env
const { default: logger } = await import("@directus/api/logger");
(await import("@directus/api/server"))
  .startServer()
  .catch((reason: unknown) => {
    logger.error(reason);
    process.exitCode = 1;
  });
```

Checking extensions dir is unnecessary, I just do it to make sure things are in place, or at least partially so.

**Reply by flevi29** [reply]
I'm dynamically importing Directus, because otherwise `esbuild` is hell bent on putting every import at the top, nullifying my env changes. You might not have to do this, if your bundler or transpiler doesn't do that, or you're writing plain javascript.

**Reply by bposman** [reply]
@flevi29 

I'm interested in your start.ts approach, what command are you using for your package.json scripts such as dev, start, etc. Finally would you mind sharing what tsconfig works for you.

I had something like this setup previously, but since the switch to ESM it doesn't really work.

**Reply by flevi29** [reply]
I only did `start`, I'm not sure how `dev` would work. Although you could probably make it work somehow, the way I designed my project, it wasn't going to work for me, so I didn't bother. It's such a clusterfu*k of unnecessary complications, recently they made you put a `package.json` for each of your plugins, so when I bundle I also have to write these `package.json`s, it's a real mess.

My start looks like this: `"start": "node --enable-source-maps --no-node-snapshot dist/start.js"`.  [Why the snapshot command line argument](https://github.com/laverdet/isolated-vm?tab=readme-ov-file#requirements).

I don't know any more what makes it work, as I didn't work on it for a long time, and it's a real mess no matter how you do it.
Only thing that really matters in your `tsconfig.json` is to set it up for ESM detailed somewhere [here](https://www.typescriptlang.org/docs/handbook/modules/reference.html#the-module-compiler-option). But I only use TypeScript for type checking, the rest is bundled with `esbuild`.
